tiledb/api/proxy/proxy.h

```diff
@@ -0,0 +1,917 @@
+};
+
+namespace {
+inline constexpr unsigned short to_index(ProxyEvent x) {
+  return static_cast<unsigned short>(x);
+}
+
+/**
+ * Number of events in the Proxy state machine
+ */
+constexpr unsigned int n_events = to_index(ProxyEvent::shutdown) + 1;
+
+}  // namespace
+
+/**
+ * The states of the state machine.
+ *
+ * The state transitions as a cartoon in ASCII art:
+ * ```
+ * nascent --> present --> destroyed
+ *  |           |    ^      ^
+ *  v           v    |      |
+ * aborted      access --> last_access
+ * ```
+ */
+enum class ProxyState {
+  /**
+   * The proxy is able to gather construction arguments. The constructor has
+   * not yet been called. The underlying object variable is `nullopt`.
+   *
+   * This is the initial state.
+   *
+   * Events
+   *   - construct. Transition to `present`. Construct the underlying object.
+   *   - access_attach. Transition to `error`.
+   *   - access_release. Transition to error.
+   *   - destroy. Transition to `aborted`.
+   *   - shutdown. Transition to `aborted`.
+   */
+  nascent = 0,
+
+  /**
+   * The constructor has been called; its result is stored in the underlying
+   * object variable. Changes to construction arguments are ignored.
+   *
+   * Events
+   *   - construct. Transition to `error`.
+   *   - access_attach. Transition to `access`. Construct an access object.
+   *   - access_release. Transition to `error`.
+   *   - destroy. Transition to `destroyed`. Destroy the underlying object.
+   *   - shutdown. Self-transition.
+   */
+  present,
+
+  /**
+   * Same as `present`, but there is a current `ProxyAccessGuard` in existence.
+   *
+   * Events
+   *   - construct. Transition to `error`.
+   *   - access_attach. Transition to `error`.
+   *   - access_release. Transition to `present`. Destroy the access object.
+   *   - destroy. Transition to `last_access`.
+   *   - shutdown. Self-transition.
+   */
+  access,
+
+  /**
+   * Same as `access`, but a destroy event has been processed. When access is
+   * released, the object will be destroyed.
+   *
+   * Events
+   *   - construct. Transition to `error`.
+   *   - access_attach. Transition to `error`.
+   *   - access_release. Transition to `destroyed`. Destroy both the underlying
+   *       and the access object.
+   *   - destroy. Self-transition.
+   *   - shutdown. Self-transition.
+   */
+  last_access,
+
+  /**
+   * An object has been destroyed and its storage deallocated. The underlying
+   * object variable is 'nullopt'. Any side-effects of the constructor may
+   * persist.
+   *
+   * This is a final state.
+   *
+   * Events
+   *   - construct. Transition to `error`.
+   *   - access_attach. Transition to `error`.
+   *   - access_release. Transition to `error`.
+   *   - destroy. Self-transition.
+   *   - shutdown. Self-transition.
+   */
+  destroyed,
+
+  /**
+   * No object was ever constructed; there are no side-effects of a constructor.
+   * The underlying object variable is 'nullopt'.
+   *
```

Comment:
It seems like in between an object being constructed and then destroyed and an object never being successfully constructed is failure during the object construction, when some side-effects could possibly happen (even writing to a log file).  Or does "any side-effects of the constructor" also include partial-but-failed construction?  In other words, aborted encompasses only failure before construction is even attempted?